core/src/main/java/de/bwaldvogel/mongo/backend/aggregation/Expression.java

```diff
@@ -1275,7 +1327,7 @@ private static Object evaluateDocumentExpression(Document expression, Document d
                 }
                 return exp.apply(expressionValue, document);
             } else {
-                result.put(expressionKey, expressionValue);
+                result.put(expressionKey, evaluate(expressionValue, document));
```

Comment:
As mentioned in the commit message for 2cfe598852bb4ba3bf91b61f955b900c6d128dc1, this change was necessary to properly evaluate the following example from [MongoDB's official documentation for $reduce](https://docs.mongodb.com/manual/reference/operator/aggregation/reduce/).

```
{
   $reduce: {
      input: [ 1, 2, 3, 4 ],
      initialValue: { sum: 5, product: 2 },
      in: {
         sum: { $add : ["$$value.sum", "$$this"] },
         product: { $multiply: [ "$$value.product", "$$this" ] }
      }
   }
}
```